File: py/bedrock/bundle.py
"""Bundle, chain, and signed-pin-update verification (spec §5.2).
Mirrors src/bundle.ts."""
from __future__ import annotations

from .errors import BedrockError
from .canon import canonical_string
from .crypto import D, verify_object
from .schema import v_bundle, v_charter, v_manifest, v_signed_pin_update
from .compiler import (
    check_bundle_semantics, check_lineage, manifest_digest_or_throw, pin_of,
)


def check_signatures(body, kind: str, signatures: list, authority: dict) -> dict:
    """Check one Detached set against an authority (§3.3 ordering rules)."""
    eligible = {k["key_id"]: k["public_key"] for k in authority["keys"]}
    ids = [s["key_id"] for s in signatures]
    if len(set(ids)) != len(ids):
        raise BedrockError("SIGNATURE_DUPLICATE", "duplicate key_id in signatures")
    for i in range(1, len(ids)):
        if ids[i - 1] >= ids[i]:
            raise BedrockError("SCHEMA", "signature entries not sorted by key_id")
    for s in signatures:
        if s["key_id"] not in eligible:
            raise BedrockError("KEY_UNKNOWN", f"key {s['key_id']} not eligible")
    for s in signatures:
        pub = eligible[s["key_id"]]
        if not verify_object(kind, body, pub, s["signature"]):
            raise BedrockError("SIGNATURE_INVALID", f"signature by {s['key_id']} invalid")
    if len(signatures) < authority["threshold"]:
        raise BedrockError("QUORUM", "below threshold")
    return {"signatures": len(signatures), "required": authority["threshold"]}


def verify_bundle(bundle: dict, root: dict, predecessors: list) -> dict:
    """Full chain verification; predecessors in any order, sorted internally."""
    b = v_bundle(bundle)
    v_charter(b["charter"])
    v_manifest(b["manifest"])
    manifest_hash = manifest_digest_or_throw(b["charter"], b["manifest"])

    chain = sorted(predecessors, key=lambda p: p["charter"]["version"])
    if b["charter"]["version"] == 1:
        if chain:
            raise BedrockError("SCHEMA", "genesis bundle with predecessors")
        if b["charter"]["previous_hash"] is not None:
            raise BedrockError("SCHEMA", "genesis previous_hash must be null")
        authority = root["bootstrap"]
    else:
        if len(chain) != b["charter"]["version"] - 1:
            raise BedrockError("HASH_MISMATCH", "incomplete predecessor chain")
        prev = None
        for p in chain:
            _verify_single(p, root, prev)
            prev = p
        pred = chain[-1]
        if pred["charter"]["version"] != b["charter"]["version"] - 1:
            raise BedrockError("VERSION_CONFLICT", "version gap")
        if b["charter"]["previous_hash"] != D("charter", pred["charter"]):
            raise BedrockError(
                "HASH_MISMATCH", "previous_hash does not name predecessor charter hash"
            )
        authority = pred["charter"]["next_authority"]

    counts = check_signatures(b["charter"], "charter", b["signatures"], authority)
    check_bundle_semantics(b["charter"], b["manifest"])
    check_lineage(b["charter"], b["manifest"], root)
    return {
        "valid": True,
        "pin": pin_of(b["charter"], manifest_hash),
        "signatures": counts["signatures"],
        "required": counts["required"],
    }


def _verify_single(bundle: dict, root: dict, predecessor):
    b = v_bundle(bundle)
    v_charter(b["charter"])
    v_manifest(b["manifest"])
    manifest_digest_or_throw(b["charter"], b["manifest"])
    if predecessor is None:
        if b["charter"]["version"] != 1 or b["charter"]["previous_hash"] is not None:
            raise BedrockError("SCHEMA", "bad genesis")
        authority = root["bootstrap"]
    else:
        if b["charter"]["version"] != predecessor["charter"]["version"] + 1:
            raise BedrockError("VERSION_CONFLICT", "version must be predecessor + 1")
        if b["charter"]["previous_hash"] != D("charter", predecessor["charter"]):
            raise BedrockError("HASH_MISMATCH", "previous_hash mismatch")
        authority = predecessor["charter"]["next_authority"]
    check_signatures(b["charter"], "charter", b["signatures"], authority)
    check_bundle_semantics(b["charter"], b["manifest"])
    check_lineage(b["charter"], b["manifest"], root)


def verify_pin_update_signatures(update: dict, authority: dict, tenant_id: str, gateway_id: str):
    u = v_signed_pin_update(update)
    if u["update"]["tenant_id"] != tenant_id or u["update"]["gateway_id"] != gateway_id:
        raise BedrockError("SCHEMA", "pin update tenant/gateway mismatch")
    check_signatures(u["update"], "pin", u["signatures"], authority)


def canonical_equal(a, b) -> bool:
    return canonical_string(a) == canonical_string(b)

File: src/bundle.ts
/**
 * Bundle, chain, and signed-pin-update verification (spec §5.2).
 */
import { BedrockError } from "./errors.js";
import { canonicalString } from "./canon.js";
import { D, verifyObject } from "./crypto.js";
import type { Authority, Bundle, Charter, Detached, Pin, RootFile, SignedPinUpdate } from "./types.js";
import { vBundle, vCharter, vManifest, vSignedPinUpdate } from "./schema.js";
import { checkBundleSemantics, checkLineage, manifestDigestOrThrow, pinOf } from "./compiler.js";

function fail(code: ConstructorParameters<typeof BedrockError>[0], msg: string): never {
  throw new BedrockError(code, msg);
}

/** Check one Detached set against an authority. Enforces §3.3 ordering rules. */
export function checkSignatures(
  body: unknown,
  kind: "charter" | "pin",
  signatures: Detached[],
  authority: Authority,
): { signatures: number; required: number } {
  const eligible = new Map(authority.keys.map((k) => [k.key_id, k.public_key]));
  // duplicate signature IDs first (SIGNATURE_DUPLICATE before set-order check)
  const ids = signatures.map((s) => s.key_id);
  if (new Set(ids).size !== ids.length) fail("SIGNATURE_DUPLICATE", "duplicate key_id in signatures");
  for (let i = 1; i < ids.length; i++) {
    if (ids[i - 1]! >= ids[i]!) fail("SCHEMA", "signature entries not sorted by key_id");
  }
  // unknown eligible-key lookup before quorum counting
  for (const s of signatures) {
    if (!eligible.has(s.key_id)) fail("KEY_UNKNOWN", `key ${s.key_id} not eligible`);
  }
  for (const s of signatures) {
    const pub = eligible.get(s.key_id)!;
    if (!verifyObject(kind, body as never, pub, s.signature)) {
      fail("SIGNATURE_INVALID", `signature by ${s.key_id} invalid`);
    }
  }
  if (signatures.length < authority.threshold) fail("QUORUM", "below threshold");
  return { signatures: signatures.length, required: authority.threshold };
}

export interface BundleVerification {
  valid: true;
  pin: Pin;
  signatures: number;
  required: number;
}

/**
 * verifyBundle(bundle, root, predecessors): full chain verification.
 * Genesis is verified against root.bootstrap; each successor under its
 * predecessor's next_authority. `predecessors` are complete bundles in
 * ascending version order.
 */
export function verifyBundle(
  bundle: Bundle,
  root: RootFile,
  predecessors: Bundle[],
): BundleVerification {
  const b = vBundle(bundle);
  vCharter(b.charter);
  vManifest(b.manifest);
  const manifestHash = manifestDigestOrThrow(b.charter, b.manifest);

  // Determine which authority must sign this version by walking the chain.
  const chain = [...predecessors].sort((a, b) => a.charter.version - b.charter.version);
  let authority: Authority;
  if (b.charter.version === 1) {
    if (chain.length !== 0) fail("SCHEMA", "genesis bundle with predecessors");
    if (b.charter.previous_hash !== null) fail("SCHEMA", "genesis previous_hash must be null");
    authority = root.bootstrap;
  } else {
    if (chain.length !== b.charter.version - 1) {
      fail("HASH_MISMATCH", "incomplete predecessor chain");
    }
    // verify the whole predecessor chain first
    let prev: Bundle | null = null;
    for (const p of chain) {
      verifySingle(p, root, prev);
      prev = p;
    }
    const pred = chain[chain.length - 1]!;
    if (pred.charter.version !== b.charter.version - 1) fail("VERSION_CONFLICT", "version gap");
    const predHash = D("charter", pred.charter);
    if (b.charter.previous_hash !== predHash) {
      fail("HASH_MISMATCH", "previous_hash does not name predecessor charter hash");
    }
    authority = pred.charter.next_authority;
  }

  const { signatures, required } = checkSignatures(b.charter, "charter", b.signatures, authority);
  checkBundleSemantics(b.charter, b.manifest);
  checkLineage(b.charter, b.manifest, root);
  return { valid: true, pin: pinOf(b.charter, manifestHash), signatures, required };
}

/** Verify one bundle's own structural/signature validity against a known predecessor. */
function verifySingle(bundle: Bundle, root: RootFile, predecessor: Bundle | null): void {
  const b = vBundle(bundle);
  vCharter(b.charter);
  vManifest(b.manifest);
  manifestDigestOrThrow(b.charter, b.manifest);
  let authority: Authority;
  if (predecessor === null) {
    if (b.charter.version !== 1 || b.charter.previous_hash !== null) {
      fail("SCHEMA", "bad genesis");
    }
    authority = root.bootstrap;
  } else {
    if (b.charter.version !== predecessor.charter.version + 1) {
      fail("VERSION_CONFLICT", "version must be predecessor + 1");
    }
    if (b.charter.previous_hash !== D("charter", predecessor.charter)) {
      fail("HASH_MISMATCH", "previous_hash mismatch");
    }
    authority = predecessor.charter.next_authority;
  }
  checkSignatures(b.charter, "charter", b.signatures, authority);
  checkBundleSemantics(b.charter, b.manifest);
  checkLineage(b.charter, b.manifest, root);
}

/**
 * Structural + signature verification of a SignedPinUpdate under `authority`.
 * CAS/expiry/head checks belong to the serving deployment (see tenant.ts).
 */
export function verifyPinUpdateSignatures(
  update: SignedPinUpdate,
  authority: Authority,
  tenantId: string,
  gatewayId: string,
): void {
  const u = vSignedPinUpdate(update);
  // foreign update fails SCHEMA before signatures are read
  if (u.update.tenant_id !== tenantId || u.update.gateway_id !== gatewayId) {
    fail("SCHEMA", "pin update tenant/gateway mismatch");
  }
  checkSignatures(u.update, "pin", u.signatures, authority);
}

export function canonicalEqual(a: unknown, b: unknown): boolean {
  return canonicalString(a as never) === canonicalString(b as never);
}
